File: parser/src/ast.cpp
#include "../include/ast.h"

namespace schematex {

// Port implementation
Port::Port(const std::string& name, Direction dir, int width)
    : name(name), direction(dir), width(width) {
}

std::string Port::toString() const {
    std::string dir_str;
    switch (direction) {
        case Direction::INPUT:
            dir_str = "input";
            break;
        case Direction::OUTPUT:
            dir_str = "output";
            break;
        case Direction::INOUT:
            dir_str = "inout";
            break;
    }
    return dir_str + " " + name + "[" + std::to_string(width) + "]";
}

// Signal implementation
Signal::Signal(const std::string& name, Type type, int width)
    : name(name), type(type), width(width) {
}

std::string Signal::toString() const {
    std::string type_str;
    switch (type) {
        case Type::WIRE:
            type_str = "wire";
            break;
        case Type::REG:
            type_str = "reg";
            break;
        case Type::INTEGER:
            type_str = "integer";
            break;
        case Type::LOGIC:
            type_str = "logic";
            break;
    }
    return type_str + " " + name + "[" + std::to_string(width) + "]";
}

// Instance implementation
Instance::Instance(const std::string& moduleName, const std::string& instanceName)
    : moduleName(moduleName), instanceName(instanceName) {
}

std::string Instance::toString() const {
    return moduleName + " " + instanceName;
}

// Assignment implementation
Assignment::Assignment(const std::string& target, const std::string& source, bool isBlocking)
    : target(target), source(source), isBlocking(isBlocking) {
}

std::string Assignment::toString() const {
    std::string op = isBlocking ? " = " : " <= ";
    return target + op + source;
}

// Block implementation
Block::Block(Type type)
    : type(type) {
}

std::string Block::toString() const {
    std::string type_str = (type == Type::SEQUENTIAL) ? "always @(posedge clk)" : "always @*";
    return type_str;
}

// Module implementation
Module::Module(const std::string& name)
    : name(name) {
}

std::string Module::toString() const {
    return "module " + name;
}

}  // namespace schematex

File: extension/src/types/ast.ts
/**
 * TypeScript interfaces for the SystemVerilog AST
 * Matches the C++ AST structure from parser/include/ast.h
 */

/**
 * Base interface for all AST nodes
 */
export interface ASTNode {
	type: string;
	[key: string]: any;
}

/**
 * Port definition (input, output, inout)
 */
export interface Port extends ASTNode {
	type: 'port';
	name: string;
	direction: 'INPUT' | 'OUTPUT' | 'INOUT';
	width: number;  // Bit width (1 for single bit, 8 for [7:0], etc.)
}

/**
 * Signal declaration (wire, reg, logic)
 */
export interface Signal extends ASTNode {
	type: 'signal';
	name: string;
	signalType: 'WIRE' | 'REG' | 'INTEGER' | 'LOGIC';
	width: number;
}

/**
 * Instance of a module
 */
export interface Instance extends ASTNode {
	type: 'instance';
	moduleName: string;
	instanceName: string;
	portConnections: Array<{
		port: string;
		signal: string;
	}>;
}

/**
 * Assignment statement
 */
export interface Assignment extends ASTNode {
	type: 'assignment';
	target: string;
	source: string;
	isBlocking: boolean;  // true for =, false for <=
}

/**
 * Block (begin...end or always block)
 */
export interface Block extends ASTNode {
	type: 'block';
	blockType: 'SEQUENTIAL' | 'COMBINATIONAL';
	statements: ASTNode[];
}

/**
 * Module definition - the root AST node
 */
export interface Module extends ASTNode {
	type: 'module';
	name: string;
	ports: Port[];
	signals: Signal[];
	instances: Instance[];
	statements: ASTNode[];  // assignments, blocks, etc.
}

/**
 * Parser error information
 */
export interface ParserError {
	/** Error message */
	message: string;
	/** Line number (1-based) */
	line: number;
	/** Column number (1-based) */
	column: number;
	/** Error type: 'parse-error', 'unsupported', 'warning', etc. */
	type: 'parse-error' | 'unsupported' | 'warning';
	/** Optional source code snippet */
	source?: string;
	/** Optional error code for categorization */
	code?: string;
}

/**
 * Complete parser output structure
 */
export interface ParserOutput {
	module?: Module;
	metadata?: {
		fileName?: string;
		parseTime?: number;
		errorCount?: number;
	};
	/** Array of parser errors and warnings */
	errors?: ParserError[];
}
